### PR TITLE
upstream: allow clusters to skip waiting on warmup for initialization

### PR DESCRIPTION
**Pull request by junr03**
Commit Message: upstream -- allow clusters to skip waiting on warmup for initialization
Additional Description: this reduces server startup time, but might cause initial 500s for requests made on unresolved clusters. Further work might be needed there if operators use this new config option.
Risk Level: low - new config option. Default uses legacy behavior.
Testing: added unit tests
Docs Changes: added.
Release Notes: updated

### CODE REVIEW

File: api/envoy/config/cluster/v3/cluster.proto [hunk 1]
```diff
@@ -876,6 +876,13 @@ message Cluster {
   // DNS resolution configuration which includes the underlying dns resolver addresses and options.
   core.v3.DnsResolutionConfig dns_resolution_config = 53;
 
+  // Optional configuration for having server init block on async DNS resolution. Only applicable for
+  // :ref:`STRICT_DNS<envoy_v3_api_enum_value_config.cluster.v3.Cluster.DiscoveryType.STRICT_DNS>`,
+  // or :ref:`LOGICAL_DNS<envoy_v3_api_enum_value_config.cluster.v3.Cluster.DiscoveryType.LOGICAL_DNS>`.
+  // If true, server startup blocks on async DNS resolution. If false, server will complete
+  // initialization whether or not DNS resolution has completed. Defaults to true.
+  google.protobuf.BoolValue wait_for_dns_on_init = 54;
```
Comment (junr03) [inline api/envoy/config/cluster/v3/cluster.proto:884]: open to a different name.

Comment (mattklein123) [reply api/envoy/config/cluster/v3/cluster.proto:884]: +1, I think this seems like a useful general option and we could just have it apply everywhere.

Comment (junr03) [reply api/envoy/config/cluster/v3/cluster.proto:884]: Great point. Will update. I can can land this PR for only dns based clusters, and if there is appetite I can do a subsequent one for other cluster types.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with some small comments.

/wait

**Comment by markdroth** [comment]
(Matt, since you're reviewing this one, I'll remove myself as API reviewer.)

File: api/envoy/config/cluster/v3/cluster.proto [hunk 2]
```diff
@@ -876,6 +876,13 @@ message Cluster {
   // DNS resolution configuration which includes the underlying dns resolver addresses and options.
   core.v3.DnsResolutionConfig dns_resolution_config = 53;
 
+  // Optional configuration for having server init block on cluster warm-up. Currently, only applicable for
```
[the remark this answers is not in the document]
Comment (junr03) [reply api/envoy/config/cluster/v3/cluster.proto]: Agreed. Let me know if this is better.

**Review by mattklein123 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**